pkg/graphviz/traceflow.go

```diff
@@ -0,0 +1,246 @@
+		return false
+	}
+	return true
+}
+
+func getSrcNodeName(tf *v1.Traceflow) string {
+	if len(tf.Source.Namespace) > 0 && len(tf.Source.Pod) > 0 {
+		return tf.Source.Namespace + "/" + tf.Source.Pod
+	}
+	return ""
+}
+
+func getDstNodeName(tf *v1.Traceflow) string {
+	if len(tf.Destination.Namespace) > 0 && len(tf.Destination.Pod) > 0 {
+		return tf.Destination.Namespace + "/" + tf.Destination.Pod
+	}
+	return ""
+}
+
+func findClusterString(graphStr string, clusterName string) (startIndex int, endIndex int) {
+	startIndex = strings.Index(graphStr, "subgraph "+clusterName)
+	if startIndex == -1 {
+		return 0, 1
+	}
+	endIndex = startIndex
+	for graphStr[endIndex] != '{' {
+		endIndex++
+	}
+	depth := 1
+	for depth > 0 {
+		endIndex++
+		if graphStr[endIndex] == '{' {
+			depth++
+		}
+		if graphStr[endIndex] == '}' {
+			depth--
+		}
+	}
+	return startIndex, endIndex + 1
+}
+
+func genOutput(g *graphviz.Graphviz, graph *cgraph.Graph, isSingleCluster bool) string {
+	var buf bytes.Buffer
+	if err := g.Render(graph, "dot", &buf); err != nil {
+		klog.Fatal(err)
+	}
+	klog.Infof("Graphz:\n%v", buf.String())
+	if err := graph.Close(); err != nil {
+		klog.Fatal(err)
+	}
+	g.Close()
+
+	str := buf.String()
+	if isSingleCluster {
+		return str
+	}
+	// Swap source and destination cluster if destination cluster appears before source cluster.
+	srcStartIdx, srcEndIdx := findClusterString(str, clusterSrcName)
+	dstStartIdx, dstEndIdx := findClusterString(str, clusterDstName)
+	if dstEndIdx <= srcStartIdx {
+		return str[:dstStartIdx] + str[srcStartIdx:srcEndIdx] + str[dstEndIdx:srcStartIdx] + str[dstStartIdx:dstEndIdx] + str[srcEndIdx:]
+	}
+	return str
+}
+
+func GenGraph(tf *v1.Traceflow) string {
+	var err error
+	g := graphviz.New()
+	graph, err := g.Graph()
+	if err != nil {
+		klog.Error(err)
+	}
+	graph.SetCenter(true)
+
+	if tf == nil || tf.Status.Phase != v1.SUCCESS || len(tf.Status.NodeSender) == 0 {
+		return genOutput(g, graph, true)
+	}
+
+	// Make the graph centered by equalizing the number of nodes on both sides.
+	var nodeNum int
+	if len(tf.Status.NodeSender) > len(tf.Status.NodeReceiver) {
+		nodeNum = len(tf.Status.NodeSender)
+	} else {
+		nodeNum = len(tf.Status.NodeReceiver)
+	}
+
+	cluster1 := graph.SubGraph(clusterSrcName, 1)
+	cluster1.SetStyle(cgraph.FilledGraphStyle)
+	// Handle single node traceflow.
+	if isSingleNodeTraceflow(tf) {
+		nodes := genSubGraph(cluster1, getSrcNodeName(tf), tf.Status.NodeSender, cgraph.ForwardDir, 0)
+		// Draw the destination pod and involved edge.
+		edgeName := fmt.Sprintf("%s_%d", cluster1.Name(), len(tf.Status.NodeSender))
+		if len(nodes) == 0 {
+			return genOutput(g, graph, true)
+		}
+		switch tf.Status.NodeSender[len(tf.Status.NodeSender)-1].Action {
+		// If the last action of the sender is FORWARDED,
+		// then the packet has been sent out by sender, implying that there is a disconnection.
+		case v1.FORWARDED:
+			lastNode, _ := graph.CreateNode("")
```

Comment:
Do you think it will be better if we can use getDstNodeName(tf) instead of "" here? I am saying this because the first example graph looks a little bit strange to me.

Reply:
> Do you think it will be better if we can use getDstNodeName(tf) instead of "" here? I am saying this because the first example graph looks a little bit strange to me.

That's a good idea.

Reply:
Modified.

Reply:
@ZhangYW18 Please click "Resolve conversation" for the comment if you think there is no need to discuss more.